File: src/integration/jvm/src/jvmMain/java/community/flock/wirespec/integration/jvm/java/HttpClientTransportation.java
package community.flock.wirespec.integration.jvm.java;

import community.flock.wirespec.java.Wirespec;

import java.net.URI;
import java.net.URLEncoder;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.util.List;
import java.util.concurrent.CompletableFuture;
import java.util.stream.Collectors;

public class HttpClientTransportation implements Wirespec.Transportation {

    private final String baseUrl;

    private final HttpClient client;

    public HttpClientTransportation(String baseUrl) {
        this(baseUrl, HttpClient.newBuilder().build());
    }

    public HttpClientTransportation(String baseUrl, HttpClient client) {
        this.baseUrl = baseUrl;
        this.client = client;
    }

    @Override
    public CompletableFuture<Wirespec.RawResponse> transport(Wirespec.RawRequest request) {

        String pathString = request.path().stream()
                .filter(it -> !it.isBlank())
                .collect(Collectors.joining("/"));

        String base = baseUrl.endsWith("/") ? baseUrl.substring(0, baseUrl.length() - 1) : baseUrl;
        String uriString = base + "/" + pathString;

        if (!request.queries().isEmpty()) {
            String queryString = request.queries().entrySet().stream()
                    .flatMap(entry -> entry.getValue().stream()
                            .map(v -> entry.getKey() + "=" + URLEncoder.encode(v, StandardCharsets.UTF_8)))
                    .collect(Collectors.joining("&"));
            uriString += "?" + queryString;
        }
        URI uri = URI.create(uriString);

        String[] headers = request.headers().entrySet().stream()
                .flatMap(entry -> entry.getValue().stream()
                        .filter(value -> !value.isBlank())
                        .flatMap(value -> List.of(entry.getKey(), value).stream()))
                .toArray(String[]::new);

        HttpRequest.Builder requestBuilder = HttpRequest.newBuilder()
                .uri(uri)
                .method(
                        request.method().toUpperCase(),
                        request.body() != null
                                ? HttpRequest.BodyPublishers.ofByteArray(request.body())
                                : HttpRequest.BodyPublishers.noBody()
                );

        if (headers.length > 0) {
            requestBuilder.headers(headers);
        }

        return client.sendAsync(requestBuilder.build(), HttpResponse.BodyHandlers.ofByteArray())
                .thenApply(response -> new Wirespec.RawResponse(
                        response.statusCode(),
                        response.headers().map(),
                        response.body()
                ));
    }
}

File: src/integration/jvm/src/jvmMain/kotlin/community/flock/wirespec/integration/jvm/HttpClientTransportation.kt
package community.flock.wirespec.integration.jvm

import community.flock.wirespec.kotlin.Wirespec
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.net.URI
import java.net.URLEncoder
import java.net.http.HttpClient
import java.net.http.HttpRequest
import java.net.http.HttpResponse

/**
 * A [Wirespec.Transportation] that executes the raw request over real HTTP using the
 * JDK [HttpClient]. A typed client edge turns a generated `Request` into a
 * [Wirespec.RawRequest]; this sends it and maps the HTTP response back into a
 * [Wirespec.RawResponse].
 */
class HttpClientTransportation(
    private val baseUrl: String,
    private val client: HttpClient = HttpClient.newHttpClient(),
) : Wirespec.Transportation {

    override suspend fun transport(request: Wirespec.RawRequest): Wirespec.RawResponse = withContext(Dispatchers.IO) {
        val response = client.send(request.toHttpRequest(), HttpResponse.BodyHandlers.ofByteArray())
        Wirespec.RawResponse(
            statusCode = response.statusCode(),
            headers = response.headers().map().mapValues { (_, v) -> v.toList() },
            body = response.body()?.takeIf { it.isNotEmpty() },
        )
    }

    private fun Wirespec.RawRequest.toHttpRequest(): HttpRequest {
        val path = path.joinToString("/") { URLEncoder.encode(it, Charsets.UTF_8) }
        val query = queries.entries
            .flatMap { (key, values) -> values.map { key to it } }
            .joinToString("&") { (k, v) ->
                "${URLEncoder.encode(k, Charsets.UTF_8)}=${URLEncoder.encode(v, Charsets.UTF_8)}"
            }
        val uri = URI.create(
            buildString {
                append(baseUrl).append("/").append(path)
                if (query.isNotEmpty()) append("?").append(query)
            },
        )

        val bodyPublisher = body
            ?.let { HttpRequest.BodyPublishers.ofByteArray(it) }
            ?: HttpRequest.BodyPublishers.noBody()

        val builder = HttpRequest.newBuilder(uri)
            .method(method, bodyPublisher)
            .header("Content-Type", "application/json")
            .header("Accept", "application/json")
        headers.forEach { (name, values) -> values.forEach { builder.header(name, it) } }
        return builder.build()
    }
}
